doc/03_reference/coverage_plan.rst

```diff
@@ -0,0 +1,299 @@
+---------------------------------------
+
+Below are lists of specific things from the microarchitecture that will be included in functional coverage.
+Each of the points listed below must be covered.
+This will be further combined using cross coverage which is described in the section below.
+
+Instructions
+^^^^^^^^^^^^
+
+Categories
+""""""""""
+Instructions can be grouped into a number of categories.
+Each category represents a different kind of execution.
+For example the ``ADD`` and ``SUB`` instructions are in the same category as they are almost identical for the microarchitecture (both read two registers and write to one, both feed operands to the ALU and take their result from it, both have the same response to interrupts etc, the only difference is the ALU operation).
+
+Instructions can be compressed or uncompressed but that isn’t factored into the instruction categories below (excepting for illegal instructions).
+The decompress occurs in the IF stage and is invisible to the ID/EX stage so isn’t relevant for instruction execution.
+A separate set of category agnostic compressed instruction behaviour is considered instead.
+
+An instruction category is sampled at the ID/EX stage (which is where all the varying behaviours actually occur).
+Some categories are just a single instruction which is named without further description.
+
+
+* ALU - All of the reg/reg reg/imm instructions that use the ALU.
+  This is any RISC-V instruction with an opcode of ``7’b0010011`` or ``7’b0110011`` (``ibex_pkg::OPCODE_OP`` and ``ibex_pkg::OPCODE_OP_IMM``) other than the ``MUL*`` and ``DIV*`` family of instructions (from RV32M).
+* Mul - Any ``MUL*`` instruction (from RV32M).
+* Div - Any ``DIV*`` instruction (from RV32M).
+* Branch - Any ``B*`` family branch instruction.
+* Jump - ``JAL``/``JALR``
+* Load - Any ``L*`` family load instruction.
+* Store - Any ``S*`` family load instruction.
+* CSRAccess - Any instruction from Zicsr.
+* EBreakDbg/EBreakExc- An ``EBREAK`` instruction that either enters debug mode (Dbg) or causes an exception (Exc).
+  Which occurs depends upon the setting of ``dcsr.ebreakm`` / ``dcsr.ebreaku`` combined with the privilege level of executed instruction.
+* ``ECALL``
+* ``MRET``
+* ``DRET``
+* ``WFI``
+* ``FENCE.I``
+* FetchError - Any instruction that saw a fetch error.
+* CompressedIllegal - Any compressed instruction with an illegal encoding.
+* UncompressedIllegal - Any uncompressed instruction with an illegal encoding.
+* CSRIllegal - Any instruction attempting a CSR access that is not allowed.
+* PrivIllegal - Illegal due to privilege level or being in/out of debug mode.
+* OtherIllegal - Any other instruction that raises an Illegal instruction exception that isn’t in the other categories.
+* None - No instruction in ID/EX stage.
+
+Stalls
+""""""
+Not all instructions can see all kinds of stalls.
+A stall category is sampled at the ID/EX stage only (as stalls in IF and WB don’t break down into categories).
```

Comment:
Are we tracking stalls in IF and WB separately? I couldn't see them mentioned elsewhere?

Reply:
Good point. For IF there's a point under `Miscellaneous` saying 'Stalled waiting an instruction fetch' I think I can move this here and rephrase so it's relative to ID/EX and hence a stall category.

For WB we have the `Mem` and `LdHz` stall categories, again relative to the ID/EX stage but we'll see these if WB is stalled (currently WB will only stall if it's awaiting a memory result).

WB stalled on its own whilst the rest of the pipeline is empty isn't a hugely interesting point and we've already got coverage for WB valid and nothing else valid to cover it. I'll add some further wording here to explain.

Reply:
I guess there's a slight difference between WB alone valid and writing back this cycle vs WB alone valid and stalled this cycle. Maybe I should expand the pipeline state section so we've got stage full & ready/full & stalled/empty and we do a big cross between those for the IF & WB stages and the ID/EX stall categories and maybe instruction categories.

Reply:
Pipeline state now has each stage with three different states, full & stalled, full & unstalled, empty (for IF instead of full & stalled, it's awaiting new instruction). There's a new cross for these states plus instruction categories.